execution-core/src/transfer/transaction.rs

```diff
@@ -0,0 +1,400 @@
+        bytes.extend((skeleton_bytes.len() as u64).to_bytes());
+        bytes.extend(skeleton_bytes);
+
+        // serialize the fee
+        bytes.extend(self.fee.to_bytes());
+
+        // serialize the deposit
+        bytes.push(u8::from(self.deposit));
+
+        // serialize the contract-call
+        match self.contract_call {
+            Some(ref call) => {
+                bytes.push(1);
+                bytes.extend(call.to_var_bytes());
+            }
+            None => {
+                bytes.push(0);
+            }
+        }
+
+        bytes
+    }
+
+    /// Deserialize the Payload from a bytes buffer.
+    ///
+    /// # Errors
+    /// Errors when the bytes are not cannonical.
+    pub fn from_slice(buf: &[u8]) -> Result<Self, BytesError> {
+        let mut buf = buf;
+
+        // deserialize the tx-skeleton
+        #[allow(clippy::cast_possible_truncation)]
+        let skeleton_len = usize::try_from(u64::from_reader(&mut buf)?)
+            .map_err(|_| BytesError::InvalidData)?;
+        let tx_skeleton = TxSkeleton::from_slice(buf)?;
+        buf = &buf[skeleton_len..];
+
+        // deserialize fee
+        let fee = Fee::from_reader(&mut buf)?;
+
+        // deserialize deposit data
+        let deposit = match u8::from_reader(&mut buf)? {
+            0 => false,
+            1 => true,
+            _ => {
+                return Err(BytesError::InvalidData);
+            }
+        };
+
+        // deserialize contract-call data
+        let contract_call = match u8::from_reader(&mut buf)? {
+            0 => None,
+            1 => Some(ContractCall::from_slice(buf)?),
+            _ => {
+                return Err(BytesError::InvalidData);
+            }
+        };
+
+        Ok(Self {
+            tx_skeleton,
+            fee,
+            deposit,
+            contract_call,
+        })
+    }
+
+    /// Return input bytes to hash the payload.
+    ///
+    /// Note: The result of this function is *only* meant to be used as an input
+    /// for hashing and *cannot* be used to deserialize the `Payload` again.
+    #[must_use]
+    pub fn to_hash_input_bytes(&self) -> Vec<u8> {
+        let mut bytes = self.tx_skeleton.to_hash_input_bytes();
+
+        bytes.push(u8::from(self.deposit));
+
+        if let Some(call) = &self.contract_call {
+            bytes.extend(call.contract);
+            bytes.extend(call.fn_name.as_bytes());
+            bytes.extend(&call.fn_args);
+        }
+
+        bytes
+    }
+
+    /// Create the `Payload`-hash to be used as an input to the
+    /// pheonix-transaction circuit.
+    #[must_use]
+    pub fn hash(&self) -> BlsScalar {
+        BlsScalar::hash_to_scalar(&self.to_hash_input_bytes())
+    }
+}
+
+/// The transaction used by the transfer contract
+#[derive(Debug, Clone, Archive, Serialize, Deserialize)]
+#[archive_attr(derive(CheckBytes))]
+pub struct Transaction {
+    payload: Payload,
+    payload_hash: BlsScalar,
+    proof: Vec<u8>,
```

Comment:
I kept these private not so much with breaking changes in mind but rather that the user will not be able to change the values of a `Transaction`. This is a protection only at the `Transaction` layer, the other new internal structs (`Payload`, `TxSkeleton`) all have public fields.

Reply:
I know that someone with bad intent will always find a way to alter a `Transaction`, so this won't protect against that. It's more with good intended players in mind so that they cannot change the `Transaction` accidentally.

Reply:
I understand this reasoning, and normally I would be sensitive to it, but given the reasoning above I don't think it's enough to make them private. As long as the contract is able to verify a transaction is valid, we should be good.